core/src/execution/datafusion/expressions/cast.rs

```diff
@@ -232,6 +232,189 @@ macro_rules! cast_int_to_int_macro {
     }};
 }
 
+// When Spark casts to Byte/Short Types, it does not cast directly to Byte/Short.
+// It casts to Int first and then to Byte/Short. Because of potential overflows in the Int cast,
+// this can cause unexpected Short/Byte cast results. Replicate this behavior.
+macro_rules! cast_float_to_int16_down {
+    (
+        $array:expr,
+        $eval_mode:expr,
+        $src_array_type:ty,
+        $dest_array_type:ty,
+        $rust_src_type:ty,
+        $rust_dest_type:ty,
+        $src_type_str:expr,
+        $dest_type_str:expr,
+        $format_str:expr
+    ) => {{
+        let cast_array = $array
+            .as_any()
+            .downcast_ref::<$src_array_type>()
+            .expect(concat!("Expected a ", stringify!($src_array_type)));
+
+        let output_array = cast_array
+            .iter()
+            .map(|value| match value {
+                Some(value) => {
+                    let is_overflow = value.is_nan() || value.abs() as i32 == std::i32::MAX;
+                    let i32_value = value as i32;
+                    match (is_overflow, $eval_mode) {
+                        (true, EvalMode::Ansi) => Err(CometError::CastOverFlow {
+                            value: format!($format_str, value).replace("e", "E"),
+                            from_type: $src_type_str.to_string(),
+                            to_type: $dest_type_str.to_string(),
+                        }),
```

Comment:
minor nit: maybe we could have a function for creating the `CastOverFlow` structs to avoid some boilerplate? See `invalid_value` for an example.

Reply:
Done.